File: socrates-web/backend/skip_reasons.py
"""Fallback skip explanations when enrichment modules don't return a reason."""

from __future__ import annotations

# IOC types each source handles (for UX when status is skipped but reason missing)
SOURCE_IOC_SUPPORT: dict[str, frozenset[str]] = {
    "VirusTotal": frozenset({"ip", "domain", "url", "hash"}),
    "Shodan": frozenset({"ip", "domain"}),
    "AbuseIPDB": frozenset({"ip", "domain"}),
    "OTX AlienVault": frozenset({"ip", "domain", "hash"}),
    "URLScan.io": frozenset({"domain", "url"}),
}

SKIP_FALLBACK: dict[str, str] = {
    "Shodan": "IP and domain lookups only",
    "AbuseIPDB": "IP and domain lookups only",
    "OTX AlienVault": "Not supported for this IOC type",
    "URLScan.io": "Domains and URLs only",
}


def skip_reason_for_source(source_display_name: str, ioc_type: str) -> str:
    """Short reason line for skipped sources (no leading 'Skipped —')."""
    supported = SOURCE_IOC_SUPPORT.get(source_display_name)
    if supported is None:
        return "Not supported for this IOC type"
    if ioc_type not in supported:
        return SKIP_FALLBACK.get(
            source_display_name,
            "Not supported for this IOC type",
        )
    return "Unavailable for this query"

File: socrates-web/frontend/src/lib/skipReasons.js
/** Fallback labels when API omits skip_reason (kept in sync with backend skip_reasons.py). */

const SOURCE_IOC_SUPPORT = {
  VirusTotal: new Set(["ip", "domain", "url", "hash"]),
  Shodan: new Set(["ip", "domain"]),
  AbuseIPDB: new Set(["ip", "domain"]),
  "OTX AlienVault": new Set(["ip", "domain", "hash"]),
  "URLScan.io": new Set(["domain", "url"]),
};

const SKIP_FALLBACK = {
  Shodan: "IP and domain lookups only",
  AbuseIPDB: "IP and domain lookups only",
  "OTX AlienVault": "Not supported for this IOC type",
  "URLScan.io": "Domains and URLs only",
};

export function skipReasonForSource(source, iocType) {
  const t = (iocType || "").toLowerCase();
  const supported = SOURCE_IOC_SUPPORT[source];
  if (!supported) return "Not supported for this IOC type";
  const effective = t && t !== "unknown" ? t : null;
  if (!effective || !supported.has(effective)) {
    return SKIP_FALLBACK[source] || "Not supported for this IOC type";
  }
  return "Unavailable for this query";
}
